CONTRIBUTING.md

```diff
@@ -0,0 +1,243 @@
+But for each change, we should think about whether they bring us closer to or further away from those goals.
+
+* Easy to understand and to extend
+  * This is an Open Source project and we embrace our community.
+    That also means that the code should be as approachable as possible.
+    That's not always easy as bytecode manipulation and instrumentation is a very complex domain.
+    Sometimes, this also conflicts with other goals, like performance.
+  * Also, a good developer documentation is critical to achieving this goal.
+    As a general guideline for code documentation,
+    document why you did something that way,
+    rather than how you did it.
+    The latter should be expressed by readable code.
+    Documenting the considered alternatives can also help others a lot to understand the problem space. 
+* Diagnosability
+  * Most issues should be resolvable when the user provides their debug logs.
+    This decreases the back-and-forth to gather additional information.
+    If an issue could not be resolved by looking at the debug logs,
+    think about if we can add the necessary information to make diagnosing similar issues easier.
+  * Try to make exception messages as descriptive as possible.
+    When this particular exception is thrown,
+    what is the user supposed to do to fix the problem?
+  * Sometimes the best debug logs are just not enough to diagnose a problem.
+    The agent code should be easily debuggable in the user's IDE.
+* Performance is a feature
+  * Find the right balance of low-level and high-level code
+  * We are especially proud of the low allocation rate of the agent.
+    This does not only the average overhead look good,
+    but also makes the performance of the agent way more predictable and stable in the higher percentiles,
+    as these are heavily influenced by GC pauses.
+  * Don't guess when it comes to performance.
+    Improvements should be verified by the benchmarks.
+    To get more insight about potential areas of improvements,
+    enable Flight Recording during the benchmarks and analyze the results in Java Mission Control.
+* Easy to install for users
+  * Installing the agent should be as easy as possible.
+    The fewer installation steps and required configuration options the user has to deal with,
+    the better.
+* Testability/Test Pyramid
+  * Most tests should be simple and easily debuggable JUnit tests.
+    In fact, the main workflow should be to tests alongside the production code.
+    You should rarely have to build the agent and attach it to a test application in order to verify your new code works as expected.
+    This makes turnarounds much faster and usually also increases the test coverage.
+    Speaking of which,
+    use code coverage tools to check that the most important branches are covered by unit tests.
+    The coverage of new code should not be lower than the current total coverage of the project.
```

Comment:
I'd temper that with "typically", and say that coverage is not a goal in and of itself and to use judgement.
(highly subjective area, so feel free to ignore me - I'm not maintaining this repo :))

Reply:
Totally agree. I recently did a change which made the coverage tool not recognize that a certain file is tested at all.